inmoose/edgepy/exactTest.py

```diff
@@ -0,0 +1,237 @@
+        the first group listed in the pair is the baseline for the comparision,
+        so if the pair is :code:`("A","B")` then the comparison is :code:`B -
+        A`, so genes with positive log-fold changes are up-regulated in group B
+        compared with group A (and vice versa for genes with negative log-fold
+        change)
+    dispersion : array_like of floats, or {"auto", "common", "trended", "tagwise"}
+        an array of dispersions or a string indicating that dispersions should
+        be taken from the data object. If floats, then can be either of length
+        one or of length equal to the number of genes. Defaults to
+        :code:`"auto"` to use the most complex dispersions found in data
+        object.
+    rejection_region : {"doubletail", "smallp", "deviance"}
+        type of rejection region for two-sided exact test.
+    big_count : int
+        count size above which asymptotic beta approximation will be used.
+    prior_count : float
+        average prior count used to shrink log-fold-changes. Larger values
+        produce more shrinkage.
+
+    Returns
+    -------
+    DGEExact
+        object containing the following components:
+
+        - :code:`table`, dataframe containing columns for the log2-fold-change
+          (:code:`"logFC"`), the average log2-counts-per-million
+          (:code:`"logCPM"`) and the two-sided *p*-value (:code:`"PValue"`).
+        - :code:`comparison`, string giving the names of the two groups being compared.
+        - :code:`genes`, dataframe containing annotation for each gene; taken
+          from :code:`self`
+    """
+    # Check input
+    if len(pair) != 2:
+        raise ValueError("Pair must be a pair!")
+    if rejection_region not in ["doubletail", "deviance", "smallp"]:
+        raise ValueError(f"Invalid value {rejection_region} for 'rejection_region'")
+
+    # Get group names
+    group = asfactor(self.samples.group)
+    levs_group = group.categories
+    if isinstance(pair[0], int) and isinstance(pair[1], int):
+        pair = levs_group[list(pair)]
+
+    if not all(np.isin(pair, levs_group)):
+        raise ValueError(
+            f"At least one element of given pair is not a group.\nGroups are {' '.join(levs_group)}"
+        )
+
+    # Get dispersion vector
+    if dispersion is None:
+        dispersion = "auto"
+    if isinstance(dispersion, str):
+        if dispersion == "common":
+            dispersion = self.common_dispersion
+        elif dispersion == "trended":
+            dispersion = self.trended_dispersion
+        elif dispersion == "tagwise":
+            dispersion = self.tagwise_dispersion
+        elif dispersion == "auto":
+            dispersion = self.getDispersion()
+        else:
+            raise ValueError(f"Invalid value {dispersion} for 'dispersion'")
+
+        if dispersion is None:
+            raise ValueError("specified dispersion not found in object")
```

Comment:
How could that happened as you already have a condition to test if the dispersion is None and input a value if so?

Reply:
Double check the `if` nesting structure.
If `dispersion` is `None`, then it is assigned the value `"auto"` (ligne 147).
Then, `str` values are dispatched, but there is no guarantee that the dispatch will fetch a non-`None` value. Typically if the dispersion have never been computed in `self`.